Source/projections/Builder/ProjectionClassBuilder.ts

```diff
@@ -0,0 +1,161 @@
+        projections.register<T>(
+            new ProjectionProcessor<T>(
+                projection,
+                client,
+                executionContext,
+                eventTypes,
+                logger
+            ), cancellation);
+    }
+
+
+
+    private tryAddAllProjectionMethods(events: EventTypeMap<[ProjectionCallback<any>, KeySelector]>, methods: OnDecoratedMethod[], eventTypes: IEventTypes, container: IContainer, logger: Logger): boolean {
+        let allMethodsValid = true;
+        for (const method of methods) {
+            const [hasEventType, eventType] = this.tryGetEventTypeFromMethod(method, eventTypes);
+
+            if (!hasEventType) {
+                allMethodsValid = false;
+                logger.warn(`Could not create projection method ${method.name} in projection ${this._projectionType.name} because it is not associated to an event type`);
+                continue;
+            }
+
+            const onMethod = this.createProjectionMethod(method, container);
+            const keySelector = method.keySelector;
+
+            if (events.has(eventType!)) {
+                allMethodsValid = false;
+                logger.warn(`Event handler ${this._projectionType.name} has multiple projection methods handling event type ${eventType}`);
+                continue;
+            }
+            events.set(eventType!, [onMethod, keySelector]);
+        }
+        return allMethodsValid;
+    }
+
+    private createProjectionMethod(method: OnDecoratedMethod, container: IContainer): ProjectionCallback<any> {
+        return (instance, event, projectionContext) => {
+            const result = method.method.call(instance, event, projectionContext);
+            if (result instanceof DeleteReadModelInstance) {
+                return result;
+            }
+            return instance;
+        };
+    }
+
+    private tryGetEventTypeFromMethod(method: OnDecoratedMethod, eventTypes: IEventTypes): [boolean, EventType | undefined] {
+        if (this.eventTypeIsId(method.eventTypeOrId)) {
+            return [
+                true,
+                new EventType(
+                    EventTypeId.from(method.eventTypeOrId),
+                    method.generation ? Generation.from(method.generation) : Generation.first)
+            ];
+        } else if (!eventTypes.hasFor(method.eventTypeOrId)) {
+            return [false, undefined];
+        } else {
+            return [true, eventTypes.getFor(method.eventTypeOrId)];
+        }
+    }
+
+    private eventTypeIsId(eventTypeOrId: Constructor<any> | EventTypeId | Guid | string): eventTypeOrId is EventTypeId | Guid | string {
+        return eventTypeOrId instanceof EventTypeId || eventTypeOrId instanceof Guid || typeof eventTypeOrId === 'string';
+    }
+
+    private ThrowIfReadModelAlreadyRegistered(types: ProjectionDecoratedType[]): void {
```

Comment:
I'm not sure I understand when this would kick in? We don't really care if a read model is used multiple times? We should really only care if a projection is registered multiple times (the Runtime would get mad).

Reply:
True, this doesn't apply anymore as the class decorated with the `@projection` becomes the read model, so for this function to throw it would be akin to having a class with the same name. Removing.